<a href="https://colab.research.google.com/github/rocx320/ImageAnalytics/blob/main/2247226_IA_LAB_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving i1.png to i1.png
Saving i2.png to i2.png
Saving i3.png to i3.png
Saving i4.png to i4.png
Saving i5.jpg to i5.jpg
Saving i6.jpg to i6.jpg
Saving i7.jpg to i7.jpg
Saving i8.jpg to i8.jpg
Saving i9.jpg to i9.jpg
Saving i10.jpg to i10.jpg


In [ ]:
import numpy as np
import pandas as pd
import cv2
from skimage.feature import greycomatrix, greycoprops
from skimage import color, img_as_ubyte

In [ ]:
# Define a function to extract features and noise-related information from an image
def extract_features(image, noise_type, noise_level):
    # Convert the image to grayscale (if not already)
    gray_image = color.rgb2gray(image)

    # Convert the grayscale image to unsigned 8-bit integer (0-255)
    gray_image = img_as_ubyte(gray_image)

    # Extract texture features using GLCM (Gray-Level Co-occurrence Matrix)
    glcm = greycomatrix(gray_image, [1], [0], symmetric=True, normed=True)
    contrast = greycoprops(glcm, 'contrast')[0, 0]
    dissimilarity = greycoprops(glcm, 'dissimilarity')[0, 0]
    homogeneity = greycoprops(glcm, 'homogeneity')[0, 0]
    energy = greycoprops(glcm, 'energy')[0, 0]
    correlation = greycoprops(glcm, 'correlation')[0, 0]

    # Extract brightness and contrast using OpenCV
    gray_image_cv2 = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    brightness = np.mean(gray_image_cv2)  # Mean pixel value is often used as an estimate of brightness
    contrast_cv2 = gray_image_cv2.std()  # Standard deviation is often used as an estimate of contrast

    return [contrast, dissimilarity, homogeneity, energy, correlation, brightness, contrast_cv2, noise_type, noise_level]

In [ ]:
# Load a list of images along with noise-related information
image_info = [
    {"path": "i1.png", "noise_type": "Gaussian", "noise_level": 0.1},
    {"path": "i2.png", "noise_type": "Gaussian", "noise_level": 0.05},
    {"path": "i3.png", "noise_type": "Gaussian", "noise_level": 0.2},
    {"path": "i4.png", "noise_type": "Gaussian", "noise_level": 0.1},
    {"path": "i5.jpg", "noise_type": "Salt-and-pepper", "noise_level": 0.04},
    {"path": "i6.jpg", "noise_type": "Salt-and-pepper", "noise_level": 0.30},
    {"path": "i7.jpg", "noise_type": "Salt-and-pepper", "noise_level": 0.4},
    {"path": "i8.jpg", "noise_type": "Speckle", "noise_level": 0.3},
    {"path": "i9.jpg", "noise_type": "Speckle", "noise_level": 0.41},
    {"path": "i10.jpg", "noise_type": "Speckle", "noise_level": 0.2},
]

In [ ]:
# Initialize an empty list to store the features for all images
feature_list = []

# Loop through each image, extract features and noise-related information, and add them to the list
for info in image_info:
    image = cv2.imread(info["path"])
    features = extract_features(image, info["noise_type"], info["noise_level"])
    feature_list.append(features)

# Convert the list of features to a Pandas DataFrame
df = pd.DataFrame(
    feature_list,
    columns=['Contrast', 'Dissimilarity', 'Homogeneity', 'Energy', 'Correlation', 'Brightness', 'Contrast_CV2', 'Noise_Type', 'Noise_Level']
)

# df is now your DataFrame with each row representing features and noise-related information for a different image, including brightness and contrast.
df.head(10)

/usr/local/lib/python3.10/dist-packages/skimage/feature/__init__.py:35: skimage_deprecation: Function ``greycomatrix`` is deprecated and will be removed in version 1.0. Use ``skimage.feature.graycomatrix`` instead.
  removed_version='1.0')
/usr/local/lib/python3.10/dist-packages/skimage/feature/__init__.py:42: skimage_deprecation: Function ``greycoprops`` is deprecated and will be removed in version 1.0. Use ``skimage.feature.graycoprops`` instead.
  removed_version='1.0')
/usr/local/lib/python3.10/dist-packages/skimage/feature/__init__.py:42: skimage_deprecation: Function ``greycoprops`` is deprecated and will be removed in version 1.0. Use ``skimage.feature.graycoprops`` instead.
  removed_version='1.0')
/usr/local/lib/python3.10/dist-packages/skimage/feature/__init__.py:42: skimage_deprecation: Function ``greycoprops`` is deprecated and will be removed in version 1.0. Use ``skimage.feature.graycoprops`` instead.
  removed_version='1.0')
/usr/local/lib/python3.10/dist-packages/skimag

Contrast  Dissimilarity  Homogeneity    Energy  Correlation  Brightness  \
0   941.414009      23.864422     0.045137  0.008768     0.748959   71.607157   
1    86.780450       5.244575     0.271814  0.020135     0.979324   84.952734   
2   202.080067       6.788843     0.238397  0.017266     0.966674  101.244449   
3   156.234263       5.158032     0.327548  0.024285     0.957093   77.988925   
4  1952.639048      32.250317     0.081598  0.032029     0.784447  126.057462   
5   175.602630       7.231262     0.259349  0.022179     0.938125   95.701712   
6  2118.545655      30.679901     0.051256  0.010743     0.668595   90.786054   
7  1705.421953      31.151877     0.038297  0.007543     0.750503  115.697897   
8   641.602540      18.755754     0.058722  0.009191     0.867775  123.185047   
9  2196.397302      37.100198     0.027687  0.006780     0.580914  130.849896   

   Contrast_CV2       Noise_Type  Noise_Level  
0     43.352483         Gaussian         0.10  
1     45.809547         Gaussian         0.05  
2     55.299299         Gaussian         0.20  
3     42.990946         Gaussian         0.10  
4     67.697397  Salt-and-pepper         0.04  
5     37.990785  Salt-and-pepper         0.30  
6     56.554349  Salt-and-pepper         0.40  
7     58.477488          Speckle         0.30  
8     49.277824          Speckle         0.41  
9     51.189266          Speckle         0.20


## Training SVM Model on the feature set

#### Testing the model performance on various test size :
- 20% Test
- 30 % Test
- 40 % Test

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

# Load your preprocessed data and labels
# X contains features (extracted from images), y contains labels (noise type and level)
from skimage.feature import greycomatrix, greycoprops
from skimage import io, color, img_as_ubyte

X = df.drop(['Noise_Type'], axis=1)
y = df.Noise_Type

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state =42)

# Create and train a machine learning model (SVM in this case)
model = SVC(kernel='linear')
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")

Accuracy: 50.00%


In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state =42)

# Create and train a machine learning model (SVM in this case)
model = SVC(kernel='linear')
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")

Accuracy: 33.33%


In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state =42)

# Create and train a machine learning model (SVM in this case)
model = SVC(kernel='linear')
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")

Accuracy: 50.00%


# Inference

#### We see that as the data sample is too small, therefore the svc is model is chosen , given its effieciency on smaller data sample

#### Also , we found the accuracy of this model to classify image on various test size in the following manner

- 50% accuracy on 20% Test Size
- 33.33% accuracy on 30% Test Size
- 50% accuracy on 40% Test Size


